File: src/promptAssets.ts
export type PromptAssetId =
  | "base"
  | "analysis"
  | "bible"
  | "outline"
  | "episode"
  | "storyboard";

export interface PromptAsset {
  id: PromptAssetId;
  name: string;
  stage: string;
  description: string;
  content: string;
}

export const DEFAULT_PROMPT_ASSETS: PromptAsset[] = [
  {
    id: "base",
    name: "总编剧系统规范",
    stage: "全管线",
    description: "控制忠实度、短剧节奏、可拍性和内容合规的基础提示词。",
    content: `You are the head writer for a vertical micro-drama. Preserve the source work's causality, character motivations, and core rules, while reorganizing it into a shootable drama rather than summarizing paragraphs.
Creative direction: establish a situation or conflict within 3–5 seconds; advance one dominant goal per episode; escalate or reverse in the middle; end on an unfinished action, identity reveal, or meaningful cost.
Expression: use conversational dialogue with subtext and minimal expository narration; concentrate locations, keep props executable, and make action shootable. Preserve character choice and emotional consequence after each payoff instead of repeating mechanical humiliation.
Safety: never glorify crime, discrimination, materialism, or degrading vulgarity. Treat minors, violence, medicine, history, and public events carefully. Do not reproduce long passages from the source.`,
  },
  {
    id: "analysis",
    name: "章节事实抽取",
    stage: "故事分析",
    description: "逐组抽取事实、事件、情绪节点和制作提示，不提前写剧本。",
    content:
      "At this stage, extract facts only and do not draft a screenplay. Every fact must trace to the input; never invent relationships or rules absent from the source.",
  },
  {
    id: "bible",
    name: "全剧事实圣经",
    stage: "连续性架构",
    description:
      "建立人物唯一身份、世界规则、时间线和道具生命周期，作为所有后续 Agent 的唯一事实源。",
    content:
      "Resolve entity identity before building the canonical story bible. Use stable IDs to connect source names, aliases, and adapted names. Rules must record state and cause; events must include location, cause, and effect; important props must track introduction, state changes, and payoff. When evidence conflicts, retain only chapter-supported facts and never patch the canon by invention.",
  },
  {
    id: "outline",
    name: "故事圣经与分集卡",
    stage: "分集规划",
    description: "合并章节证据，生成全剧定位、集目标、反转和结尾卡点。",
    content:
      "Use the story bible as the sole factual source when planning episodes. Keep one core conflict and introduce new information in every episode. Adjacent episodes must not repeat the same humiliation, breakup, or threat. Establish a visual conflict immediately, escalate in the middle, and end with a reason to continue. Declare entry state, exit state, and causal transition from the previous episode. Determine scene count from actual action beats, location changes, and emotional rhythm.",
  },
  {
    id: "episode",
    name: "逐集可拍成稿",
    stage: "剧本生成",
    description: "依据分集卡和上一集连续性生成场次、动作与对白。",
    content:
      "Follow the story bible, episode contract, and previous-episode continuity exactly. Keep action concise, specific, and executable. Dialogue must be conversational and contain subtext; avoid using dialogue to explain the entire backstory. Every scene must change a character, information, prop, or relationship state.",
  },
  {
    id: "storyboard",
    name: "分镜与多媒体制作包",
    stage: "分镜制作",
    description: "控制镜头拆分、构图、运镜、声音设计和图像提示词，但不改变已批准的故事。",
    content:
      "Convert only the approved screenplay into production-ready shots. Preserve facts, characters, locations, outcomes, screen direction, props, wardrobe, and visual anchors. Specify shot size, camera movement, composition, performance action, sound, continuity, and production notes. Image and negative prompts must be English for broad image-model compatibility.",
  },
];

const LEGACY_DEFAULT_HASHES: Partial<Record<PromptAssetId, string>> = {
  base: "c50c670d",
  analysis: "1a67053",
  bible: "d94b8f4b",
  outline: "9e9766b5",
  episode: "ba8dda74",
};

function contentHash(value: string) {
  let hash = 2166136261;
  for (const character of value) {
    hash ^= character.codePointAt(0) || 0;
    hash = Math.imul(hash, 16777619);
  }
  return (hash >>> 0).toString(16);
}

export function loadPromptAssets(): PromptAsset[] {
  try {
    const stored = JSON.parse(
      localStorage.getItem("scriptforge.prompt-assets.v1") || "[]",
    ) as Partial<PromptAsset>[];
    return DEFAULT_PROMPT_ASSETS.map((fallback) => {
      const saved = stored.find((item) => item.id === fallback.id);
      const legacyDefault =
        typeof saved?.content === "string" &&
        contentHash(saved.content) === LEGACY_DEFAULT_HASHES[fallback.id];
      return {
        ...fallback,
        ...(legacyDefault ? {} : saved || {}),
        id: fallback.id,
      };
    });
  } catch {
    return DEFAULT_PROMPT_ASSETS;
  }
}

export function promptContent(
  assets: PromptAsset[] | undefined,
  id: PromptAssetId,
) {
  return (
    assets?.find((asset) => asset.id === id)?.content ||
    DEFAULT_PROMPT_ASSETS.find((asset) => asset.id === id)?.content ||
    ""
  );
}

File: macos/Sources/ScriptForgeMac/PromptAssets.swift
import Foundation
import CryptoKit

enum PromptAssets {
    static let version = "mac-prompts-v4-en"

    static let defaults: [PromptAsset] = [
        PromptAsset(
            id: "story-evidence",
            title: "Story Evidence Extraction",
            scope: "Chapter analysis and the evidence layer for Book Analysis",
            influence: "Controls which characters, events, causal links, and foreshadowing elements are retained from the source; it does not directly set screenplay style.",
            instruction: """
            Extract only facts explicitly supported by the source. Every fact must retain its chapter evidence ID. Distinguish completed events, character motivations, world rules, and unresolved foreshadowing. Do not add information absent from the source.
            """,
            isDefault: true,
            updatedAt: Date(timeIntervalSince1970: 0)
        ),
        PromptAsset(
            id: "character-naming",
            title: "Lightweight Character Renaming",
            scope: "Flash-model naming call after import",
            influence: "Controls new screenplay character names. Manually edited names have the highest priority and must not be changed by later model calls.",
            instruction: """
            Generate natural Chinese names that fit each character's personality and the conventions of Chinese short-form drama. Each name must contain two to four Chinese characters, be unique, differ from the source name, and never use placeholders such as "Character 8" or "Male Lead 1."
            """,
            isDefault: true,
            updatedAt: Date(timeIntervalSince1970: 0)
        ),
        PromptAsset(
            id: "story-bible",
            title: "Story Bible and Continuity",
            scope: "Series facts, relationships, timeline, and prop continuity",
            influence: "Constrains character identity, relationships, ability rules, location changes, and foreshadowing payoffs across all episodes; it is the canonical source for continuity decisions.",
            instruction: """
            Organize chapter evidence into an actionable story bible. Reconcile character identities and aliases, state the causes and limits of world rules, and build a causally ordered timeline. Never present unsupported material as established fact.
            """,
            isDefault: true,
            updatedAt: Date(timeIntervalSince1970: 0)
        ),
        PromptAsset(
            id: "episode-planning",
            title: "Chinese Vertical-Drama Episode Planning",
            scope: "Series-wide episode contracts and per-episode scene recommendations",
            influence: "Controls each episode's new event, opening hook, reversal, closing cliffhanger, and dynamic scene count; it does not draft complete dialogue.",
            instruction: """
            Advance one central conflict per episode while adding new information or causing a new consequence. Show a visible conflict within the first five seconds and end the final five to eight seconds on an unfinished action or information gap. Within the first three episodes, establish the situation, a humiliation or crisis, and an ability or truth reversal. Never repeat the same conflict across consecutive episodes. Let the story determine scene count; a 60-second episode normally uses one to three scenes.
            """,
            isDefault: true,
            updatedAt: Date(timeIntervalSince1970: 0)
        ),
        PromptAsset(
            id: "episode-drafting",
            title: "Shootable Episode Draft",
            scope: "Per-episode scene action and dialogue",
            influence: "Directly controls pacing, visible action, dialogue density, and the closing cliffhanger in the final screenplay.",
            instruction: """
            Draft a production-ready vertical micro-drama episode for the Chinese market. Every action must be visible, shootable, and give performers a playable reaction. Keep dialogue short and adversarial; never restate an action the audience has just seen. A 60-second episode normally contains 12–18 dialogue lines, preferably no more than 18 Chinese characters per line. Do not split scenes merely to increase the count. End at the exact moment an action, discovery, or choice occurs.
            """,
            isDefault: true,
            updatedAt: Date(timeIntervalSince1970: 0)
        ),
        PromptAsset(
            id: "quality-gate",
            title: "Evidence-Grounded Quality Gate",
            scope: "Episode review, series audit, issue ledger, and targeted repair",
            influence: "Determines whether a draft passes and identifies repair locations; it must not rewrite valid content merely to increase a score.",
            instruction: """
            Use evidence to check source fidelity, character continuity, motivation, time and location transitions, conflict progression, opening and closing hooks, and shootability. Report only issues with a precise location. During repair, address only blocker and major issues while preserving scenes and dialogue that already work.
            """,
            isDefault: true,
            updatedAt: Date(timeIntervalSince1970: 0)
        ),
        PromptAsset(
            id: "storyboard-production",
            title: "Storyboard and Multimedia Production Package",
            scope: "Approved screenplay to shots, keyframes, sound, and continuity notes",
            influence: "Controls shot breakdown, composition, camera movement, sound design, and image prompts without changing the approved story.",
            instruction: """
            Break the approved screenplay into executable shots. Every shot must specify shot size, camera movement, vertical composition, visible action, dialogue or narration, sound, continuity, and production notes. Total shot duration must match the episode target. Each keyframe prompt must repeat stable character visual anchors and specify 9:16. Do not add characters, events, locations, prop abilities, or outcomes absent from the screenplay.
            """,
            isDefault: true,
            updatedAt: Date(timeIntervalSince1970: 0)
        ),
        PromptAsset(
            id: "book-analysis",
            title: "Six-Section Book Analysis",
            scope: "Structure, characters, commercial payoffs, style, techniques, and adaptation guidance",
            influence: "Controls the depth and evidence standard of the analysis report; it does not directly modify the screenplay draft.",
            instruction: """
            The book analysis must cover six sections: content overview, structure and pacing, character system, commercial highlights, language and style, and reusable techniques. Every judgment must cite chapter evidence IDs. Clearly distinguish source facts, analytical inferences, and adaptation recommendations.
            """,
            isDefault: true,
            updatedAt: Date(timeIntervalSince1970: 0)
        ),
    ]

    // Hashes allow old built-in defaults to migrate without shipping their obsolete prompt text.
    private static let legacyDefaultHashes: [String: String] = [
        "story-evidence": "f5b356263bab21ae72baad20086e8b0a21fc95a1658b2339c62f7bc2e5996b1e",
        "character-naming": "4eefac4b6513b21e62ebd235eb18e08d8929c86369ed06973b8e681a065bee39",
        "story-bible": "10b34c08e82fadd8d5ad6d0592ea19fa48e3f2da2fab3f7f2e519aecc7a65b49",
        "episode-planning": "d1009c1a1d53d3f53941b2311b51ece1226e34ec8465deebb2b26678ee03d82b",
        "episode-drafting": "fbfe098b3cab1fa8f6bbb4a7fccae2fe337fbdb7872e3f7b474a5d3618b276f9",
        "quality-gate": "833a0f4add1b081c4a02130f156e0444b2fdb87382d22a6e2793e8702762d3b6",
        "storyboard-production": "f2821351fc758be5c217554c325850d6fc8f05114f74912e4388312f7848e323",
        "book-analysis": "e99298dbfcc41ce056a976b0c816e7869109c77270d2e81487bccd0df689cf7f",
    ]

    static func mergedInstruction(_ ids: [String], assets: [PromptAsset]) -> String {
        ids.compactMap { id in assets.first(where: { $0.id == id })?.instruction }
            .joined(separator: "\n\n")
    }

    static func reset(asset: PromptAsset) -> PromptAsset {
        defaults.first(where: { $0.id == asset.id }) ?? asset
    }

    static func isLegacyDefault(_ asset: PromptAsset) -> Bool {
        guard let expected = legacyDefaultHashes[asset.id] else { return false }
        let normalized = asset.instruction.trimmingCharacters(in: .whitespacesAndNewlines)
        let digest = SHA256.hash(data: Data(normalized.utf8)).map { String(format: "%02x", $0) }.joined()
        return digest == expected
    }
}

final class PromptAssetRepository {
    private let fileURL: URL
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    init(baseURL: URL? = nil) {
        let root = baseURL ?? FileManager.default.urls(
            for: .applicationSupportDirectory,
            in: .userDomainMask
        ).first!.appendingPathComponent("ScriptForge", isDirectory: true)
        fileURL = root.appendingPathComponent("prompt-assets.json")
        encoder = JSONEncoder.scriptForge
        decoder = JSONDecoder.scriptForge
    }

    func load() -> [PromptAsset] {
        guard
            let data = try? Data(contentsOf: fileURL),
            let saved = try? decoder.decode([PromptAsset].self, from: data)
        else { return PromptAssets.defaults }
        let merged = PromptAssets.defaults.map { defaultAsset in
            guard let savedAsset = saved.first(where: { $0.id == defaultAsset.id }) else {
                return defaultAsset
            }
            guard !savedAsset.isDefault, !PromptAssets.isLegacyDefault(savedAsset) else {
                return defaultAsset
            }
            return PromptAsset(
                id: defaultAsset.id,
                title: defaultAsset.title,
                scope: defaultAsset.scope,
                influence: defaultAsset.influence,
                instruction: savedAsset.instruction,
                isDefault: false,
                updatedAt: savedAsset.updatedAt
            )
        }
        if merged != saved { try? save(merged) }
        return merged
    }

    func save(_ assets: [PromptAsset]) throws {
        try FileManager.default.createDirectory(
            at: fileURL.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        try encoder.encode(assets).write(to: fileURL, options: .atomic)
    }
}

extension JSONEncoder {
    static var scriptForge: JSONEncoder {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
        return encoder
    }
}

extension JSONDecoder {
    static var scriptForge: JSONDecoder {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }
}
